rfd/0196-introducing-a-component-system.md

```diff
@@ -0,0 +1,518 @@
+```
+
+### Our building blocks aren't customisable enough
+
+Similar to above, due to us not using `styled-system` enough, it's common to see patterns like:
+
+```tsx
+const PreviewWrapper = styled(Box)`
+  border-radius: ${p => p.theme.radii[3]}px;
+  box-shadow: ${p => p.theme.boxShadow[1]};
+  transform: var(--feature-preview-scale);
+  background-color: ${p => p.theme.colors.levels.surface};
+`;
+```
+
+Due to the `Box` component not accepting all props that developers expect. This could just be:
+
+```tsx
+<Box
+  borderRadius="lg"
+  boxShadow="md"
+  transform="var(--feature-preview-scale)"
+  backgroundColor="levels.surface"
+>
+  ...
+</Box>
+```
+
+The first code example takes longer to write, as it has to be defined outside the component and then referred to
+in the component. A name has to be thought up, and to use values from the theme, a longer syntax of `${p => p.theme.whatever.is.needed[2]}` 
+has to be used, instead of simply using `value="whatever.is.needed.2"`. This also takes longer for a developer coming back to the code
+to understand where the styling is.
+
+### Our components are custom-made
+
+We currently have to reinvent the wheel a lot of the time when creating components. This is fine for some components, but
+we're missing accessibility and features that are built into pre-built components. Our tooltip component lacks the ability
+to know if the tooltip will be displayed off the screen, and our modal component doesn't have the ability to close with the 
+`Esc` key. 
+
+Using pre-built components from Chakra UI would enhance developer velocity, as they do not have to worry
+about implementing these features.
+
+## Advantages of using Chakra UI
+
+Chakra UI is a React component system that has been around for over 5 years, with 38k stars on GitHub and a large
+community around it. There are Chakra versions of other 3rd-party libraries we use, such as `react-select`, that allow
+for consistent theming and styling across the codebase.
+
+Chakra is a framework that is very similar to how we currently write components, making it an easier fit into the codebase
+without having to relearn a new way of writing components. 
+
+### Pre-built components
+
+Chakra UI has [a lot of pre-built components](https://chakra-ui.com/docs/components/concepts/overview) that are ready to
+use, and offer better implementations than our current custom-made components (Tooltip, Modal, etc.).
+
+As we introduce Chakra UI components into the Teleport codebase, the recipes will need to be altered to match the
+Teleport designs. This should be pretty straightforward, and most of the time be spacing/color changes.
+
+### Easy styling
+
+As mentioned above, Chakra adds the ability to create recipes, which allow the definition of reusable components to
+stay in
+one place, and be reused throughout the codebase. Recipes also make defining dark mode variants much easier.
```

Comment:
Regarding dark mode variants, I feel like because we have to maintain a custom theme for that one customer, this doesn't really apply to us because we should be thinking in terms of themes, not dark/light mode, no?

Reply:
That seems like a better overall model for this anyway. Currently we have three separate themes, but it might make sense for customers to want to implement dark/light modes for their themes as well.